### ISSUE TITLE
Reinit logic clarification

### ISSUE THREAD

**Issue by tomleavy** [mlswg/mls-protocol#591]
The goal of the reinit proposal is to trigger a tear down / restart of a group with the same members but different parameters such as a version upgrade, or a cipher suite change.
 
1. Changing version - version is a field within key package, and this locks a key package to a specific version, even if capabilities allow for multiple versions. Should that field be removed, and instead you just look at capabilities? It seems that this would change with #575 since leaf nodes will not be pinned to a specific version, just a cipher suite. After 575 gets merged one could upgrade versions using a new tree that copies the current non-empty leaves so long as each leaf has the upgraded version in their capabilities set. That specific tree building process is not mentioned, it is just sort of assumed in context.
 
2. Changing cipher suite - In the case you want to change cipher suite, complexity somewhat increases. You will need to create a new tree by grabbing new key packages for each participant, and because the cipher suite has changed, all of the credentials have also changed. It is mentioned that "A group may be reinitialized by creating a new group with the same membership and different parameters", but this does not have a statement that has a SHOULD or MUST associated with it. For example, in my 5000 member group 1 member does not support the cipher suite that we are migrating away from. As the committer, I'm assuming that I should keep everyone in the group as best effort, but have the option to filter out problematic members that are blocking a migration due to them being offline for long periods of time or having outdated software. Due to the credential change, enforcing a MUST might be difficult since equality between two x509 credentials is not straight forward for example.
 
To summarize, I propose some changes
 
1. Remove version from key package, and rely on capabilities instead, this may now be irrelevant due to #575, but the version and capabilities.versions values seem duplicative
 
2. Add some language in reinitialization that clarifies the flow for upgrading just version vs changing cipher suite, where changing cipher suite explicitly requires you acquire new key packages for each member.
 
3. Clarify that a person issuing the welcome based on a reinit proposal SHOULD include all existing members

**Comment by bifurcation** [comment]
1. IMO those fields have quite separate purposes.  Keep in mind that a different version could have a different KeyPackage format.  So `key_package.version` says "this KeyPackage is formatted according to this version", while the capabilities says what other versions I support.
2. With that in mind, you need to grab new KeyPackages for everyone regardless of whether you're changing version, ciphersuite, or both.  If this isn't clear, we should certainly clarify it.
3. If you're worried about people not supporting the new stuff, you could maintain the hard semantic just by Removing those participants that don't support the new stuff before you ReInit.

**Comment by tomleavy** [comment]
#1 - Fair enough that seems reasonable

#2 - I don't think it's necessarily clear right now, it's more of an assumption based on an understanding of the protocol overall. I'll try to come up with something there to give more of a step by step explanation of the flow. 

#3 - If this is a hard requirement should the doc say MUST here? If so, I'm trying to think through what equality means in terms of membership when in the event the cipher suite changes, all the credentials will also change.

**Comment by bifurcation** [comment]
On (3) -- I would be OK with saying that the new members MUST be the same as the old members.  The notion of "same" is a little squishy, given ambiguities around identifiers, but it seems like it will probably be clear in practice.  W.r.t. removing people to make an upgrade possible, when the application decides to upgrade and how the application it meets the same-member criterion is up to the application.
